src/proxy/MapProxy.ts

```diff
@@ -100,119 +99,192 @@ type EntryEventHandler = (key: Data, value: Data, oldValue: Data, mergingValue:
-        const predicateData = this.toData(predicate);
-        return this.encodeInvokeOnRandomTarget(MapAggregateWithPredicateCodec, aggregatorData, predicateData)
-            .then((clientMessage) => {
-                const response = MapAggregateWithPredicateCodec.decodeResponse(clientMessage);
-                return this.toObject(response);
-            });
+        let aggregatorData, predicateData;
+        try {
+            aggregatorData = this.toData(aggregator);
+            predicateData = this.toData(predicate);
+        } catch (e) {
+            if (e instanceof SchemaNotReplicatedError) {
+                return this.registerSchema(e.schema, e.clazz).then(() => this.aggregateWithPredicate(aggregator, predicate));
+            }
+            throw e;
+        }
+        return this.encodeInvokeOnRandomTarget(MapAggregateWithPredicateCodec, (clientMessage) => {
+            const response = MapAggregateWithPredicateCodec.decodeResponse(clientMessage);
+            return this.toObject(response);
+        }, aggregatorData, predicateData);
     }
 
-    executeOnKeys(keys: K[], entryProcessor: IdentifiedDataSerializable | Portable): Promise<any[]> {
+    executeOnKeys(keys: K[], entryProcessor: any): Promise<any[]> {
         assertNotNull(keys);
         assertArray(keys);
         if (keys.length === 0) {
             return Promise.resolve([]);
         } else {
-            const toObject = this.toObject.bind(this);
-            const keysData = SerializationUtil.serializeList(this.toData.bind(this), keys);
-            const proData = this.toData(entryProcessor);
-            return this.encodeInvokeOnRandomTarget(MapExecuteOnKeysCodec, proData, keysData)
-                .then<Array<[K, V]>>((clientMessage) => {
-                    const response = MapExecuteOnKeysCodec.decodeResponse(clientMessage);
-                    return SerializationUtil.deserializeEntryList(toObject, response);
-                });
+            let keysData: Data[];
+            let proData: Data;
+            try {
+                keysData = this.serializeList(keys);
+                proData = this.toData(entryProcessor);
+            } catch (e) {
+                if (e instanceof SchemaNotReplicatedError) {
+                    return this.registerSchema(e.schema, e.clazz).then(() => this.executeOnKeys(keys, entryProcessor));
+                }
+                throw e;
+            }
+            return this.encodeInvokeOnRandomTarget(MapExecuteOnKeysCodec, (clientMessage) => {
+                const response = MapExecuteOnKeysCodec.decodeResponse(clientMessage);
+                return this.deserializeEntryList(response);
+            }, proData, keysData);
         }
     }
 
-    executeOnKey(key: K, entryProcessor: IdentifiedDataSerializable | Portable): Promise<V> {
+    executeOnKey(key: K, entryProcessor: any): Promise<V> {
         assertNotNull(key);
         assertNotNull(entryProcessor);
-        const keyData = this.toData(key);
```

Comment:
Ok let me explain. We thought doing this in BaseProxy:

```js
protected toData(object: any, partitioningStrategy?: any): Promise<Data> {
    try {
        return Promise.resolve(this.serializationService.toData(object, partitioningStrategy));
    } catch (e) {
        if (e instanceof SchemaNotReplicatedError) {
            return this.registerSchema(e.schema, e.clazz).then(() => this.toData(object, partitioningStrategy));
        }
        throw e;
    }
}
```

Since fetching the schema and returning the data afterwards is an async operation,  `return this.registerSchema(e.schema, e.clazz).then(() => this.toData(object, partitioningStrategy));` returns a `Promise<Data>`. The return type should be uniform so we need to return a Promise in the happy case as well. We can use the immediately resolving promise Promise.resolve() but it is problematic. It has a performance cost like %15 compared to just returning the Data synchronously. It is because promises work with the event loop.

Recently, @yuce  proposed another way of putting the try catch blocks into a single place. Something like:

```js
protected toData(object: any, partitioningStrategy?: any): Promise<[Data | null, SchemaNotReplicatedError | null]> {
    try {
        return [this.serializationService.toData(object, partitioningStrategy), null];
    } catch (e) {
        if (e instanceof SchemaNotReplicatedError) {
            return [null, e];
        }
        throw e;
    }
}
```

The toData method in BaseProxy will have a try/catch block and return an error value which is nullable SchemaNotReplicatedError type. If it is not null we will understand that we need to register a schema:


```js
const [data, error] = this.toData(element);
if (error !== null) {
    return this.registerSchema(error.schema, error.clazz).then(() => this.add(element));
}
return this.encodeInvoke(ListAddCodec, ListAddCodec.decodeResponse, data);
```

Reply:
Forgot to mention: We won't implement Yuce's proposal for now because it will take some time. However, I want to implement it in the future.